src/extensions/sort.rs

```diff
@@ -0,0 +1,155 @@
+//! Adds support for the IMAP SORT extension specificed in [RFC
+//! 5464](https://tools.ietf.org/html/rfc5256#section-3).
+//!
+//! The SORT command is a variant of SEARCH with sorting semantics for
+//! the results. There are two arguments before the searching
+//! criteria argument: a parenthesized list of sort criteria, and the
+//! searching charset.
+
+use std::{borrow::Cow, fmt};
+
+pub(crate) struct SortCriteria<'c>(pub(crate) &'c [SortCriterion<'c>]);
+
+impl<'c> fmt::Display for SortCriteria<'c> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.0.is_empty() {
+            write!(f, "")
+        } else {
+            let criteria: Vec<String> = self.0.iter().map(|c| c.to_string()).collect();
+            write!(f, "({})", criteria.join(" "))
+        }
+    }
+}
+
+/// Message sorting preferences used for [`Session::sort`] and [`Session::uid_sort`].
+///
+/// References:
+/// - [IMAP ref](https://tools.ietf.org/html/rfc5256#ref-IMAP)
+/// - [Address formal syntax](https://tools.ietf.org/html/rfc3501#section-9)
+/// - [Address spec](https://tools.ietf.org/html/rfc2822#section-3.4.1)
```

Comment:
Let's write this in a more helpful way:
```suggestion
/// Any sorting criterion that refers to an address (`From`, `To`, etc.) sorts according to the "addr-mailbox" of the indicated address. You can find the formal syntax for addr-mailbox [in the IMAP spec](https://tools.ietf.org/html/rfc3501#section-9), and a more detailed discussion of the relevant semantics [in RFC 2822](https://tools.ietf.org/html/rfc2822#section-3.4.1). Essentially, the address refers _either_ to the name of the contact _or_ to its local-part (the left part of the email address, before the `@`).
```

After adding ^, we can simplify all the documentation strings for the individual variants below.

Reply:
:heavy_check_mark: Done